roles/container-engine/containerd/templates/config.toml.j2

```diff
@@ -8,7 +8,19 @@ oom_score = {{ containerd_oom_score }}
   max_send_message_size = {{ containerd_grpc_max_send_message_size }}
 
 [debug]
```

Comment:
Unless I'm mistaken, the debug socket is not enabled if this section is not present.
IMO, this whole section should be guarded with something like `containerd_enabled_debug_socket`, defaulting to false.
Production setup should presumably not have a debug socket lying around.

Reply:
The debug log level is not the same thing that the normal log level, is it not ?

I know that section is already present. What I mean is that it should not be enabled by default, as kubespray by default target production use cases, and having a debug socket enabled in production is not optimal.

Reply:
It set the log level for everything, not just debug ones.
Same for the log format inside it.

Do you still want to disable this section entirely with a `containerd_enabled_debug` bool ? And no if defined inside it ?

See: https://github.com/containerd/containerd/blob/83031836b2cf55637d7abf847b17134c51b38e53/cmd/containerd/command/main.go#L348

It takes the config.Debug.Level if no cli params is used, cli flags have higher priorities but we do not use them in kubespray.

```go
func setLogLevel(context *cli.Context, config *srvconfig.Config) error {
	l := context.GlobalString("log-level")
	if l == "" {
		l = config.Debug.Level
	}
	if l != "" {
		return log.SetLevel(l)
	}
	return nil
}
```

Config struct is found [here](https://github.com/containerd/containerd/blob/v1.7.16/services/server/config/config.go) and the debug section looks like this:

```go
type Debug struct {
	Address string `toml:"address"`
	UID     int    `toml:"uid"`
	GID     int    `toml:"gid"`
	Level   string `toml:"level"`
	// Format represents the logging format. Supported values are 'text' and 'json'.
	Format string `toml:"format"`
}
```

Reply:
Wow, this is so weird (or I totally misunderstood something). Does that mean there is no way to change the log level from the config without enabling the debug socket ?
I mean, I assumed that the "normal" log stream (which I presume goes to standard output) and the "debug" log stream (to the debug socket) where different, but are they in fact the same ?

I've tried to check containerd documentation but couldn't find explanation about this.

-----
That said, I still thinks we should not enable the debug socket in production, (unless the debug socket is an unfortunate naming and it's in fact required ?). Since we handle the generation of the systemd unit for containerd we could inject the log level as flags... :thinking: 

That does seems a bit weird from containerd though, I feel like I'm missing something.

Reply:
Having the [debug] section in the toml config does not mean we activate the debug socket.
It is only activated `if config.Debug.Address != ""`, you can see the code here: https://github.com/containerd/containerd/blob/v1.7.16/cmd/containerd/command/main.go#L232

```go
		if config.Debug.Address != "" {
			var l net.Listener
			if isLocalAddress(config.Debug.Address) {
				if l, err = sys.GetLocalListener(config.Debug.Address, config.Debug.UID, config.Debug.GID); err != nil {
					return fmt.Errorf("failed to get listener for debug endpoint: %w", err)
				}
			} else {
				if l, err = net.Listen("tcp", config.Debug.Address); err != nil {
					return fmt.Errorf("failed to get listener for debug endpoint: %w", err)
				}
			}
			serve(ctx, l, server.ServeDebug)
		}
```

It is exactly the same for the metrics one: https://github.com/containerd/containerd/blob/v1.7.16/cmd/containerd/command/main.go#L245

```go
		if config.Metrics.Address != "" {
			l, err := net.Listen("tcp", config.Metrics.Address)
			if err != nil {
				return fmt.Errorf("failed to get listener for metrics endpoint: %w", err)
			}
			serve(ctx, l, server.ServeMetrics)
		}
```

Reply:
Ok, I think I got it, thanks for bearing with me (https://github.com/containerd/containerd/blob/main/docs/man/containerd-config.toml.5.md is not super helpful here !).

So, defining the address is what activate the debug socket, got it.
But according to the link above, it has a default value of `/run/containerd/debug.sock`, so it would be activated by default, is that correct ?

What I would do in that case is have in our defaults/main.yml something like that.
```yml
containerd_debug_enable: false
contaienrd_debug_address: '/run/containerd/debug.sock
# the rest of the value with default matching containerd 
```
And in the template something likg
```toml
[debug]
     address = "{{ containerd_debug_enable | ternary(containerd_debug_address, '') }}"
     # the rest of the values
 ```
 wdyt ?

Reply:
The doc is misleading, it says it is enabled by default but it is not.

Here the config from a newly created node in my cluster
```toml
version = 2
root = "/var/lib/containerd"
state = "/run/containerd"
oom_score = 0

[grpc]
  max_recv_message_size = 16777216
  max_send_message_size = 16777216

[debug]
  level = "info"

[metrics]
  address = ""
  grpc_histogram = false

[plugins]
  [plugins."io.containerd.grpc.v1.cri"]
    sandbox_image = "registry.k8s.io/pause:3.9"
    max_container_log_line_size = -1
    enable_unprivileged_ports = false
    enable_unprivileged_icmp = false
  [plugins."io.containerd.grpc.v1.cri".containerd]
    default_runtime_name = "runc"
    snapshotter = "overlayfs"
  
    [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]
      [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
        runtime_type = "io.containerd.runc.v2"
        runtime_engine = ""
        runtime_root = ""
        base_runtime_spec = "/etc/containerd/cri-base.json"
    
          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
            systemdCgroup = true
            binaryName = "/opt/bin/runc"
```

No debug.sock is setup nor present on the node, it seems to default to "", we can check with `/opt/bin/containerd config dump` which gives us

```toml
[debug]
  address = ""
  format = ""
  gid = 0
  level = "info"
  uid = 0
```

We can setup it explicitely like you said, it may be less confusing for us and users.

Reply:
I pushed my modifications